firedrake/bcs.py

```diff
@@ -295,3 +315,72 @@ def homogenize(bc):
     except TypeError:
         # not iterable
         return DirichletBC(bc.function_space(), 0, bc.sub_domain)
+
+
+class FormBC(BCObject):
+    r'''Implementation of a form boundary condition.
+
+    :arg V: the :class:`.FunctionSpace` on which the boundary condition
+        should be applied.
+    :arg g: the boundary condition values. This can be a :class:`.Function` on
+        ``V``, a :class:`.Constant`, an :class:`.Expression`, an
+        iterable of literal constants (converted to an
+        :class:`.Expression`), or a literal constant which can be
+        pointwise evaluated at the nodes of
+        ``V``. :class:`.Expression`\s are projected onto ``V`` if it
+        does not support pointwise evaluation.
+    :arg sub_domain: the integer id(s) of the boundary region over which the
+        boundary condition should be applied. The string "on_boundary" may be used
+        to indicate all of the boundaries of the domain. In the case of extrusion
+        the ``top`` and ``bottom`` strings are used to flag the bcs application on
+        the top and bottom boundaries of the extruded mesh respectively.
+    :arg method: the method for determining boundary nodes. The default is
+        "topological", indicating that nodes topologically associated with a
+        boundary facet will be included. The alternative value is "geometric",
+        which indicates that nodes associated with basis functions which do not
+        vanish on the boundary will be included. This can be used to impose
+        strong boundary conditions on DG spaces, or no-slip conditions on HDiv spaces.
+    '''
+
+    def __init__(self, V, eq, sub_domain, J=None, Jp=None, method="topological"):
+        super().__init__(V, sub_domain, method="topological")
+        self.formeq = eq
+        self.F = None
+        self.J = J
+        self.Jp = Jp
+        # f (= F, J or Jp) is set right before calling "asesmble(...),"
+        # telling the function which one of residual, Jacobian, or
+        # Preconditionar is to be assembled.
+        self.f = None
```

Comment:
This feels like bad design, because it is stateful, and likely to be the cause of difficult to find bugs.